src/mcfret.c

```diff
@@ -0,0 +1,366 @@
+          
+         time_now=log_time(time_now,log);
+         fclose(log);
+    }/*Closing the loop over samples*/
+    
+    /*The calculation is finished, lets write output*/
+    log=fopen("NISE.log","a");
+    fprintf(log,"Finished Calculating Response!\n");
+    fprintf(log,"Writing to file!\n");  
+    fclose(log);
+
+    if (non->cluster!=-1){
+        printf("Of %d samples %d belonged to cluster %d.\n",samples,Ncl,non->cluster);
+        if (samples==0){ /*Avoid dividing by zero*/ 
+            samples=1;
+        }
+    }
+
+    fclose(H_traj);
+    if (non->cluster!=-1){
+        fclose(Cfile);
+    }
+
+  /* Save time domain response */
+    absorption_matrix=fopen("TD_absorption_matrix.dat","w");
+    emission_matrix=fopen("TD_emission_matrix.dat","w");
+    for (t1=0;t1<non->tmax1;t1+=non->dt1){
+        fprintf(absorption_matrix,"%f %e %e\n",t1*non->deltat,re_S_Abs[t1]/samples,im_S_Abs[t1]/samples);
+        fprintf(emission_matrix, "%f %e %e\n",t1*non->deltat,re_S_Emi[t1]/samples,im_S_Emi[t1]/samples);
+    }
+    fclose(absorption_matrix);
+    fclose(emission_matrix);
+    
+    
+    /*Free the memory*/
+    
+    free(re_S_Abs);
+    free(im_S_Abs);
+    free(re_S_Emi);
+    free(im_S_Emi);
+
+    
+    free(vecr);	
+    free(veci);
+
+    free(dens_mat);
+
+    
+}
+void density_matrix(float *density_matrix, float *Hamiltonian_i,t_non *non){
+  /*This function will create a density matrix where every term is weighted with a Boltzmann weight*/
+  int index,N;
+  float *H,*e,*c2,*c1;
+  float *cnr;
+  /*float *crr;*/
+  N=non->singles;
+  H=(float *)calloc(N*N,sizeof(float));
+  e=(float *)calloc(N,sizeof(float));
+  c2=(float *)calloc(N,sizeof(float));
+  /*c1=(float *)calloc(N,sizeof(float));*/
+  cnr=(float *)calloc(N*N,sizeof(float));
+  /*crr=(float *)calloc(N*N,sizeof(float));*/
+  int a,b,c;
+  float kBT=non->temperature*0.695; // Kelvin to cm-1
+  float Q,iQ;
+
+  /*Build Hamiltonian*/
+  for (a=0;a<N;a++){
+      H[a+N*a]=Hamiltonian_i[a+N*a-(a*(a+1))/2]; /*Diagonal*/
+      for (b=a+1;b<N;b++){
+          H[a+N*b]=Hamiltonian_i[b+N*a-(a*(a+1))/2];
+          H[b+N*a]=Hamiltonian_i[b+N*a-(a*(a+1))/2];
+      }
+  }
+  diagonalizeLPD(H,e,N);
+ 
+  /*Exponentiate [U=exp(-H/kBT)]*/
+  for (a=0;a<N;a++){
+      c2[a]=exp(-e[a]/kBT);
+      Q=Q+c2[a];
+  }
+  iQ=1.0/Q;
+
+  /*Transform to site basis*/ 
+  for (a=0;a<N;a++){
+      for (b=0;b<N;b++){
+          cnr[b+a*N]+=H[b+a*N]*c2[b]*iQ;
+      }
+  }  
+  for (a=0;a<N;a++){
+      for (b=0;b<N;b++){
+          for (c=0;c<N;c++){
+              density_matrix[a+c*N]+=H[b+a*N]*cnr[b+c*N];
+          }
+      }
+  }
+  
+  free(H);
+  free(c2);
+  free(c1);
+  free(e);
+  /*free(crr);*/
+  free(cnr);
+  return;
+}
+
+void mcfret_autodetect(t_non *non, float treshold);
+
+void mcfret_response_function(float* Hamiltonian, float *re_S_1,float *im_S_1,int t1,t_non *non,float *cr,float *ci, float *density_matrix, int emission){
```

Comment:
Now this function works with a toggle parameter `emission`. I would like to propose to create two separate functions. One for emission and one for absorption. For example: `mcfret_emission_function()` and `mcfret_absorption_function()`.